### ISSUE TITLE
Ensure RabbitMQ messages are processed despite microservices failures

### ISSUE THREAD

**Issue by yoomlam** [department-of-veterans-affairs/abd-vro#565]
Ensure RabbitMQ messages are processed despite microservices failures by testing scenarios where the microservice fails to process a request (i.e., RabbitMQ message) and ensuring the message is redelivered to a functional microservice. 

From https://www.rabbitmq.com/tutorials/amqp-concepts.html#message-acknowledge, I believe we want to use `basic.ack`:
> * After the application sends back an acknowledgement (using the basic.ack method).
> ...
> If a consumer dies without sending an acknowledgement, the broker will redeliver it to another consumer or, if none are available at the time, the broker will wait until at least one consumer is registered for the same queue before attempting redelivery.

https://www.rabbitmq.com/tutorials/amqp-concepts.html#acknowledgements

Also consider exponential backoff: https://www.brianstorti.com/rabbitmq-exponential-backoff/

**Comment by nanotone** [comment]
From a code analysis, currently every Python service is using automatic acknowledgements. That is, in every call to `channel.basic_consume`, the `auto_ack` parameter is set to True.

**Comment by yoomlam** [comment]
Data point: As I'm working on something else, I restarted the `svc-lighthouse-api` Java microservice only, then resubmitted a request and got an error and the `app` container:
```
2022-12-10 19:14:50.854  INFO 1 --- [nio-8080-exec-3] gov.va.vro.controller.VroController      : Getting full health assessment for claim 1234 and veteran icn 9000682
2022-12-10 19:14:51.052  INFO 1 --- [nio-8080-exec-3] g.v.v.s.p.camel.SlipClaimSubmitRouter    : Routing to rabbitmq:claim-submit-exchange?queue=claim-submit&routingKey=code.7101&requestTimeout=60000.
2022-12-10 19:15:51.729  WARN 1 --- [ubmit-exchange]] o.a.c.c.r.reply.ReplyManagerSupport      : Timeout occurred after 60000 millis waiting for reply message with correlationID [Camel-4C99BE31F6CD64D-0000000000000004] on destination amq.gen-7E7WYi_Irq4JYC399i7BYQ. Setting ExchangeTimedOutException on (MessageId: 4C99BE31F6CD64D-0000000000000003 on ExchangeId: 4C99BE31F6CD64D-0000000000000003) and continue routing.
2022-12-10 19:15:51.784 ERROR 1 --- [ubmit-exchange]] o.a.c.p.e.DefaultErrorHandler            : Failed delivery for (MessageId: 4C99BE31F6CD64D-0000000000000003 on ExchangeId: 4C99BE31F6CD64D-0000000000000003). Exhausted after delivery attempt: 1 caught: org.apache.camel.ExchangeTimedOutException: The OUT message was not received within: 60000 millis due reply message with correlationID: Camel-4C99BE31F6CD64D-0000000000000004 not received on destination: amq.gen-7E7WYi_Irq4JYC399i7BYQ. Exchange[4C99BE31F6CD64D-0000000000000003]
Message History (source location and message history is disabled)
---------------------------------------------------------------------------------------------------------------------------------------
Source                                   ID                             Processor                                          Elapsed (ms)
                                         claim-submit-full/claim-submit from[direct://claim-submit-full]                      559279571
	...
                                         claim-submit-full/routingSlip4 routingSlip[bean[SlipClaimSubmitRouter method:rout            0
Stacktrace
---------------------------------------------------------------------------------------------------------------------------------------
org.apache.camel.ExchangeTimedOutException: The OUT message was not received within: 60000 millis due reply message with correlationID: Camel-4C99BE31F6CD64D-0000000000000004 not received on destination: amq.gen-7E7WYi_Irq4JYC399i7BYQ. Exchange[4C99BE31F6CD64D-0000000000000003]
	at org.apache.camel.component.rabbitmq.reply.ReplyManagerSupport.processReply(ReplyManagerSupport.java:174) ~[camel-rabbitmq-3.19.0.jar!/:3.19.0]
	at org.apache.camel.component.rabbitmq.reply.TemporaryQueueReplyHandler.onTimeout(TemporaryQueueReplyHandler.java:67) ~[camel-rabbitmq-3.19.0.jar!/:3.19.0]

...
2022-12-10 19:15:51.796 ERROR 1 --- [nio-8080-exec-3] gov.va.vro.controller.VroController      : Error in full health assessment
org.apache.camel.CamelExecutionException: Exception occurred during execution on the exchange: Exchange[4C99BE31F6CD64D-0000000000000003]
	at org.apache.camel.CamelExecutionException.wrapCamelExecutionException(CamelExecutionException.java:45) ~[camel-api-3.19.0.jar!/:3.19.0]
	at org.apache.camel.support.ExchangeHelper.extractResultBody(ExchangeHelper.java:660) ~[camel-support-3.19.0.jar!/:3.19.0]
...
Caused by: org.apache.camel.ExchangeTimedOutException: The OUT message was not received within: 60000 millis due reply message with correlationID: Camel-4C99BE31F6CD64D-0000000000000004 not received on destination: amq.gen-7E7WYi_Irq4JYC399i7BYQ. Exchange[4C99BE31F6CD64D-0000000000000003]
	at org.apache.camel.component.rabbitmq.reply.ReplyManagerSupport.processReply(ReplyManagerSupport.java:174) ~[camel-rabbitmq-3.19.0.jar!/:3.19.0]
	at org.apache.camel.component.rabbitmq.reply.TemporaryQueueReplyHandler.onTimeout(TemporaryQueueReplyHandler.java:67) ~[camel-rabbitmq-3.19.0.jar!/:3.19.0]
	at org.apache.camel.component.rabbitmq.reply.CorrelationTimeoutMap.listener(CorrelationTimeoutMap.java:49) ~[camel-rabbitmq-3.19.0.jar!/:3.19.0]
...
```

Experimenting, I restarted `app` container only and resubmitted API request -- it worked as normal.
So the `app` container may need updating/configuring to reconnect to (previously) failed message deliveries. I did not investigate further.

**Comment by nanotone** [comment]
Re: localhost:15672 -- Thanks for this helpful tip. The Features are per-queue, and auto-ack is a per-consumer option; multiple consumers could subscribe to the same queue with different ack settings.

I just confirmed that if a Python microservice intentionally raises an uncaught exception right after receiving a message, the message does not get redelivered. This matches my understanding of the use of `auto_ack` as described above.

For the time being, I actually recommend accepting this shortcoming, because:
- Currently all microservices are on the hook for a "synchronous" response (it's async of course, but with an expectation of a quick response) so the `app` HTTP request handler just returns an error after 20 seconds of waiting.
- All Python code is wrapped in an exception handler that forwards a fail-fast error back to the request handler. I assume Java microservice code is similar.
- When auto-ack is off, RabbitMQ's timeout before attempting re-delivery is **30 minutes** (which might explain your observation above). While this value is configurable, the feature seems to be more for ensuring that important long-running tasks get completed.

To me, there isn't a strong benefit to covering the scenarios where the network fails to deliver a message _after_ RabbitMQ gets its TCP ack, or where the microservice fails at the OS or machine level before sending back a response. We should just rely on the app's 20 second timeout for these. But when we have microservices working on tasks where completion can't conceivably be expected by an HTTP handler, then we should use explicit acks.

**Comment by yoomlam** [comment]
@nanotone's recommendation is acceptable. However, I'd like to propose some microservice conventions (and possibly a new VRO feature). First some points:
1. The reason that auto-ack is acceptable for API requests is because the `app` container has an API request timeout, so the originator of the request will get a timeout error regardless of the type of error in the microservice. This potentially hides the actual error and can make debugging difficult.
2. If another container is the originator of the request to the microservice, that container is responsible for ensuring requests are not dropped due to a microservice failure.

Proposal:
- Add the convention that _EVERY_ microservice must provide a response within T seconds, where T can be different for each request type.
- If a response is not provided in time, a timeout occurs for that microservice request and (_something_) triggers the request to be resubmitted (potentially for another instance of the microservice to consume).
  - As long as the reply queue (I forgot the exact RabbitMQ header name -- maybe `correspondence-queue`) is used, the response for the resubmitted request should go to the original requester (and not the _something_ that resubmitted the message).
  - "_something_" could be a VRO feature that checks for timed-out requests, tracked in something like Redis. This implies all microservice requests be logged in Redis, along with its timeout value T. Further discussion on the _something_ is needed.

Benefits of this proposal: VRO is not reliant on RabbitMQ-specific features (e.g., auto-ack).

**Comment by yoomlam** [comment]
After some reading on [RabbitMQ's acknowledgements](https://www.rabbitmq.com/confirms.html), I have some relevant notes that present an alternative proposal to the above:
- There are 2 reasons for acks: 1) consumer received the message or 2) consumer successfully processed the message

> Depending on the acknowledgement mode used, RabbitMQ can consider a message to be successfully delivered either immediately after it is sent out (written to a TCP socket) or when an explicit ("manual") client acknowledgement is received. Manually sent acknowledgements can be positive or negative and use one of the following protocol methods:
> - basic.ack is used for positive acknowledgements
> - basic.nack is used for negative acknowledgements (note: this is a [RabbitMQ extension to AMQP 0-9-1](https://www.rabbitmq.com/nack.html))
> - basic.reject is used for negative acknowledgements but has one limitation compared to basic.nack
> ...
> Positive acknowledgements simply instruct RabbitMQ to record a message as delivered and can be discarded. Negative acknowledgements with basic.reject have the same effect.

If the consumer gracefully fails, it can send a basic.nack so that the message is requeued for another consumer. If the consumer is disconnected, we'd have to rely on a timeout or disconnection detection -- see Disconnections section below.

> In automatic acknowledgement mode, a message is considered to be successfully delivered immediately after it is sent. This mode trades off higher throughput (as long as the consumers can keep up) for reduced safety of delivery and consumer processing. This mode is often referred to as "fire-and-forget". Unlike with manual acknowledgement model, if consumers's TCP connection or channel is closed before successful delivery, the message sent by the server will be lost. Therefore, automatic message acknowledgement should be considered unsafe and not suitable for all workloads.

We could use manual acks, and treat acks as "processing is complete" then immediately send the response (if any) to the requester.


We'd probably want `requeue=true`
> Such deliveries can be discarded or dead-lettered or requeued by the broker. This behaviour is controlled by the requeue field. When the field is set to true, the broker will requeue the delivery (or multiple deliveries, as will be explained shortly) with the specified delivery tag. Alternatively, when this field is set to false, the message will be routed to a [Dead Letter Exchange](https://www.rabbitmq.com/dlx.html) if it is configured, otherwise it will be discarded.

We'd want `multiple=false`:
> If multiple was set to false, deliveries 5, 6, and 7 would still be unacknowledged.

> manual acknowledgements from clients are also inherently asynchronous in nature. So there's a sliding window of delivery tags that are unacknowledged. Developers would often prefer to cap the size of this window to avoid the unbounded buffer problem on the consumer end. This is done by setting a "prefetch count" value using the basic.qos method. 

Maybe set `basic.qos` to 2x the number of consumers typically listening on the queue?
> The QoS prefetch setting has no effect on messages fetched using the basic.get ("pull API"), even in manual confirmation mode.

Not sure if our microservices use `basic.get`.

> When manual acknowledgements are used, any delivery (message) that was not acked is automatically requeued when the channel (or connection) on which the delivery happened is closed. This includes TCP connection loss by clients, consumer application (process) failures, and channel-level protocol exceptions
> ... Redeliveries will have a special boolean property, redeliver, set to true by RabbitMQ. For first time deliveries it will be set to false.

## Disconnections

From https://www.rabbitmq.com/heartbeats.html,
> The heartbeat timeout value defines after what period of time the peer TCP connection should be considered unreachable (down) by RabbitMQ and client libraries.
> ... The value is in seconds, and default value suggested by RabbitMQ is 60.

There's lots of details that I glossed over for now.

## Publisher Confirms

Publisher confirms seems unnecessary. There will be other data safety measures at the API request boundary to ensure all API requests are completely processed.

**Comment by nanotone** [comment]
Note that RabbitMQ only supports a single consumer timeout for all consumers, so we can't rely on it for varying values of T based on message type. (Not that we were strongly inclined to do so, but it was brought up in engineering huddle as an implementation possibility)

https://www.rabbitmq.com/consumers.html#acknowledgement-timeout

**Comment by yoomlam** [comment]
Taking a step back...
If all our microservices (message consumers) always sends back a response, then it is up to the requester (message producer) to handle any errors (i.e., no response) and no special MQ acknowledge features are needed. So we need conventions on how requesters handle errors, such as resubmitting the request (for a functioning consumer).

**Comment by sethdarragile6** [comment]
> Data point: As I'm working on something else, I restarted the `svc-lighthouse-api` Java microservice only, then resubmitted a request and got an error and the `app` container:
> 
> ```
> 2022-12-10 19:14:50.854  INFO 1 --- [nio-8080-exec-3] gov.va.vro.controller.VroController      : Getting full health assessment for claim 1234 and veteran icn 9000682
> 2022-12-10 19:14:51.052  INFO 1 --- [nio-8080-exec-3] g.v.v.s.p.camel.SlipClaimSubmitRouter    : Routing to rabbitmq:claim-submit-exchange?queue=claim-submit&routingKey=code.7101&requestTimeout=60000.
> 2022-12-10 19:15:51.729  WARN 1 --- [ubmit-exchange]] o.a.c.c.r.reply.ReplyManagerSupport      : Timeout occurred after 60000 millis waiting for reply message with correlationID [Camel-4C99BE31F6CD64D-0000000000000004] on destination amq.gen-7E7WYi_Irq4JYC399i7BYQ. Setting ExchangeTimedOutException on (MessageId: 4C99BE31F6CD64D-0000000000000003 on ExchangeId: 4C99BE31F6CD64D-0000000000000003) and continue routing.
> 2022-12-10 19:15:51.784 ERROR 1 --- [ubmit-exchange]] o.a.c.p.e.DefaultErrorHandler            : Failed delivery for (MessageId: 4C99BE31F6CD64D-0000000000000003 on ExchangeId: 4C99BE31F6CD64D-0000000000000003). Exhausted after delivery attempt: 1 caught: org.apache.camel.ExchangeTimedOutException: The OUT message was not received within: 60000 millis due reply message with correlationID: Camel-4C99BE31F6CD64D-0000000000000004 not received on destination: amq.gen-7E7WYi_Irq4JYC399i7BYQ. Exchange[4C99BE31F6CD64D-0000000000000003]
> Message History (source location and message history is disabled)
> ---------------------------------------------------------------------------------------------------------------------------------------
> Source                                   ID                             Processor                                          Elapsed (ms)
>                                          claim-submit-full/claim-submit from[direct://claim-submit-full]                      559279571
> 	...
>                                          claim-submit-full/routingSlip4 routingSlip[bean[SlipClaimSubmitRouter method:rout            0
> Stacktrace
> ---------------------------------------------------------------------------------------------------------------------------------------
> org.apache.camel.ExchangeTimedOutException: The OUT message was not received within: 60000 millis due reply message with correlationID: Camel-4C99BE31F6CD64D-0000000000000004 not received on destination: amq.gen-7E7WYi_Irq4JYC399i7BYQ. Exchange[4C99BE31F6CD64D-0000000000000003]
> 	at org.apache.camel.component.rabbitmq.reply.ReplyManagerSupport.processReply(ReplyManagerSupport.java:174) ~[camel-rabbitmq-3.19.0.jar!/:3.19.0]
> 	at org.apache.camel.component.rabbitmq.reply.TemporaryQueueReplyHandler.onTimeout(TemporaryQueueReplyHandler.java:67) ~[camel-rabbitmq-3.19.0.jar!/:3.19.0]
> 
> ...
> 2022-12-10 19:15:51.796 ERROR 1 --- [nio-8080-exec-3] gov.va.vro.controller.VroController      : Error in full health assessment
> org.apache.camel.CamelExecutionException: Exception occurred during execution on the exchange: Exchange[4C99BE31F6CD64D-0000000000000003]
> 	at org.apache.camel.CamelExecutionException.wrapCamelExecutionException(CamelExecutionException.java:45) ~[camel-api-3.19.0.jar!/:3.19.0]
> 	at org.apache.camel.support.ExchangeHelper.extractResultBody(ExchangeHelper.java:660) ~[camel-support-3.19.0.jar!/:3.19.0]
> ...
> Caused by: org.apache.camel.ExchangeTimedOutException: The OUT message was not received within: 60000 millis due reply message with correlationID: Camel-4C99BE31F6CD64D-0000000000000004 not received on destination: amq.gen-7E7WYi_Irq4JYC399i7BYQ. Exchange[4C99BE31F6CD64D-0000000000000003]
> 	at org.apache.camel.component.rabbitmq.reply.ReplyManagerSupport.processReply(ReplyManagerSupport.java:174) ~[camel-rabbitmq-3.19.0.jar!/:3.19.0]
> 	at org.apache.camel.component.rabbitmq.reply.TemporaryQueueReplyHandler.onTimeout(TemporaryQueueReplyHandler.java:67) ~[camel-rabbitmq-3.19.0.jar!/:3.19.0]
> 	at org.apache.camel.component.rabbitmq.reply.CorrelationTimeoutMap.listener(CorrelationTimeoutMap.java:49) ~[camel-rabbitmq-3.19.0.jar!/:3.19.0]
> ...
> ```
> 
> Experimenting, I restarted `app` container only and resubmitted API request -- it worked as normal. So the `app` container may need updating/configuring to reconnect to (previously) failed message deliveries. I did not investigate further.

I was able to reproduce this as well, I'll write up a ticket and inform Amida. @yoomlam 

I think I found another problem with resiliency when restarting rabbitmq, hopefully this isn't _totally_ redundant with the rabbitmq discussion above. 

1. When stopping the rabbitmq container, swagger call (/v1/full-health-data-assessment with "diagnosticCode": "7101") just spins forever, and the logs in each container keep filling up with errors about not finding rabbitmq
2. After bringing rabbitmq back, resubmitting the swagger call continues to spin forever. 
3. Only after restarting the app-1 container, resubmitting the swagger call succeeds

I'll write up this ticket as well

**Comment by sethdarragile6** [comment]
☝ I meant to add that comment to #527 , not this ticket. Moving it to there...

**Comment by yoomlam** [comment]
> So we need conventions on how requesters handle errors, such as resubmitting the request (to a functioning consumer).

Let's close this ticket and write a new ticket to: Document conventions on how requesters handle errors, such as resubmitting the request with various options like exponential backoff. We should borrow and link to articles that provide common retry patterns.

**Comment by nanotone** [comment]
@yoomlam Should this new ticket be an addendum to #900, or take the place of (and close) #900?

**Comment by yoomlam** [comment]
I didn't know about that ticket, which is sufficient. It think this ticket can be closed.

**Comment by nanotone** [comment]
I'll transfer over some of this verbiage to that ticket, so add some specificity to the acceptance criteria.
